pkg/controller/nodeipam/ipam/multinetwork_cloud_cidr_allocator.go

```diff
@@ -0,0 +1,77 @@
+package ipam
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	compute "google.golang.org/api/compute/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	networkv1 "k8s.io/cloud-provider-gcp/crd/apis/network/v1"
+	"k8s.io/klog/v2"
+)
+
+// PerformMultiNetworkCIDRAllocation allots pod CIDRs for all the networks that a node is connected to. It handles IPv6 only for default-network for now.
+func (ca *cloudCIDRAllocator) PerformMultiNetworkCIDRAllocation(node *v1.Node, interfaces []*compute.NetworkInterface) (defaultNwCIDRs []string, northInterfaces networkv1.NorthInterfacesAnnotation, additionalNodeNetworks networkv1.MultiNetworkAnnotation, err error) {
+	k8sNetworksList, err := ca.networkClient.NetworkingV1().Networks().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("error fetching networks: %v", err)
+	}
+	k8sNetworks := make([]networkv1.Network, 0)
+	// ignore networks that are under deletion.
+	for _, network := range k8sNetworksList.Items {
+		if network.ObjectMeta.DeletionTimestamp.IsZero() {
+			k8sNetworks = append(k8sNetworks, network)
+		}
+	}
+	gkeNwParamsClient := ca.networkClient.NetworkingV1().GKENetworkParams()
+	// Fetch the GKENetworkParams for every k8s-network object.
+	// Match the fetched GKENetworkParams object with the interfaces on the node
+	// to build the per-network north-interface and node-network annotations useful for IPAM.
+	for _, inf := range interfaces {
+		rangeNameAliasIPMap := map[string]*compute.AliasIpRange{}
+		for _, ipRange := range inf.AliasIpRanges {
+			rangeNameAliasIPMap[ipRange.SubnetworkRangeName] = ipRange
+		}
+		for _, network := range k8sNetworks {
+			klog.V(4).Infof("allotting pod cidrs for network %s", network.Name)
+			gnp, err := gkeNwParamsClient.Get(context.TODO(), network.Spec.ParametersRef.Name, metav1.GetOptions{})
+			if err != nil {
+				return nil, nil, nil, err
+			}
+			if resourceName(inf.Network) != resourceName(gnp.Spec.VPC) || resourceName(inf.Subnetwork) != resourceName(gnp.Spec.VPCSubnet) {
```

Comment:
You don't need 'resourceName' call for gnp.Sepc.VPC right?

Reply:
Just having an additional filter in case users give a range name in the wrong format